### ISSUE TITLE
Function with decorator + required kwargs is not type checked properly

### ISSUE THREAD

**Issue by dekim24** [python/mypy#16685]
**Bug Report**

If you wrap a function containing required keyword arguments with a decorator, type checking doesn't behave as one might expect - if you call the function and it's missing those kwargs (and/or contains unexpected kwargs), the mypy type check succeeds.

**To Reproduce**

```python
from functools import lru_cache

@lru_cache(maxsize=128)
def my_function(required_param: str) -> None:
    pass

my_function(color="blue", speed="fast")
```

Run `mypy <filename>` - observe that it succeeds
Run `python <filename>` - observe error

Note that the actual code I was testing in our codebase uses a different decorator. I just picked a random one that is publicly available

**Expected Behavior**

mypy returns a typecheck error because `required_param` is not set

**Actual Behavior**

`Success: no issues found in 1 source file`

if I try to execute this code, python raises the expected `TypeError: my_function() got an unexpected keyword argument 'color'`

if I remove the decorator mypy correctly raises a typecheck error.

**Your Environment**

- Mypy version used: 1.7.1 (latest)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None that would impact this case
- Python version used: 3.11.6

**Comment by erictraut** [comment]
This is a duplicate of https://github.com/python/mypy/issues/16261. For additional details, see https://github.com/python/mypy/issues/5107#issuecomment-1355954910.

This isn't a problem with decorators in general. It's specific to `lru_cache`, and it's because of the way `lru_cache` is defined in the typeshed stubs.

**Comment by dekim24** [comment]
> This is a duplicate of #16261. For additional details, see [#5107 (comment)](https://github.com/python/mypy/issues/5107#issuecomment-1355954910).
> 
> This isn't a problem with decorators in general. It's specific to `lru_cache`, and it's because of the way `lru_cache` is defined in the typeshed stubs.

I agree it's not a mypy bug, but as I mentioned in the issue description, the actual decorator we implemented in our code is not `lru_cache`. Actually the underlying issue is that while our function is type hinted, the decorator was not type hinted which causes the entire function signature to get ignored.
